src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp

```diff
@@ -0,0 +1,239 @@
+      "targets.");
+
+    return address_of_exprt{target, pointer_type(target.type())};
+  }
+  UNREACHABLE;
+}
+
+/// \brief Generates instructions to conditionally snapshot the value
+/// of `target_pointer` into `target_snapshot_var`.
+///
+/// ```
+/// DECL target_valid_var : bool
+/// DECL target_snapshot_var : <target_pointer.type()>
+/// ASSIGN target_valid_var := <all_dereferences_valid(target_pointer)>
+/// ASSIGN target_snapshot_var := NULL
+/// IF !target_valid_var GOTO skip_target
+/// ASSIGN target_snapshot_var := target_pointer;
+/// skip_target: SKIP
+/// ```
+///
+static void snapshot_if_valid(
+  const symbol_exprt &target_valid_var,
+  const symbol_exprt &target_snapshot_var,
+  const exprt &target_pointer,
+  goto_programt &dest,
+  // context parameters
+  const namespacet &ns)
+{
+  dest.add(goto_programt::make_decl(target_valid_var));
+
+  dest.add(goto_programt::make_decl(target_snapshot_var));
+
+  dest.add(goto_programt::make_assignment(
+    target_valid_var, all_dereferences_are_valid(target_pointer, ns)));
+
+  dest.add(goto_programt::make_assignment(
+    target_snapshot_var,
+    null_pointer_exprt{to_pointer_type(target_pointer.type())}));
+
+  goto_programt skip_program;
+  const auto skip_target = skip_program.add(goto_programt::make_skip());
+
+  dest.add(goto_programt::make_goto(skip_target, not_exprt{target_valid_var}));
+
+  dest.add(goto_programt::make_assignment(target_snapshot_var, target_pointer));
+
+  dest.destructive_append(skip_program);
+}
+
+/// \brief Generates instructions to conditionally havoc
+/// the given `target_snapshot_var`.
+///
+/// Generates these instructions if target is a `__CPROVER_POINTER_OBJECT(...)`:
+///
+/// ```
+/// IF !target_valid_var GOTO skip_target
+/// OTHER havoc_object(target_snapshot_var)
+/// skip_target: SKIP
+/// DEAD target_valid_var
+/// DEAD target_snapshot_var
+/// ```
+///
+/// And generate these instructions otherwise:
+///
+/// ```
+/// IF !target_valid_var GOTO skip_target
+/// ASSIGN *target_snapshot_var = nondet()
+/// skip_target: SKIP
+/// DEAD target_valid_var
+/// DEAD target_snapshot_var
+/// ```
+/// Adds a special comment on the havoc instructions in order to trace back
+/// the havoc to the replaced function.
+static void havoc_if_valid(
+  const symbol_exprt &target_valid_var,
+  const symbol_exprt &target_snapshot_var,
+  const exprt &target,
+  const irep_idt &tracking_comment,
+  goto_programt &dest,
+  // context parameters
+  const source_locationt &source_location,
+  const namespacet &ns)
+{
+  goto_programt skip_program;
+  const auto skip_target =
+    skip_program.add(goto_programt::make_skip(source_location));
+
+  dest.add(goto_programt::make_goto(
+    skip_target, not_exprt{target_valid_var}, source_location));
+
+  if(target.id() == ID_pointer_object)
+  {
+    // OTHER __CPROVER_havoc_object(target_snapshot_var)
+    codet code(ID_havoc_object);
+    code.add_to_operands(target_snapshot_var);
```

Comment:
`codet code(ID_havoc_object, {target_snapshot_var});` would avoid the two-step construction.

Reply:
fixed